test/pylib/resource_gather.py

```diff
@@ -0,0 +1,195 @@
+        self.test_id: int = self.sqlite_writer.write_row_if_not_exist(
+            Test(
+                architecture=platform.machine(),
+                directory=test.suite.name,
+                mode=test.mode,
+                run_id=test.id,
+                test_name=test.shortname
+            ),
+            TESTS_TABLE)
+
+    def make_cgroup(self):
+        os.makedirs(self.cgroup_path, exist_ok=True)
+
+    def get_test_metrics(self) -> Metric:
+        test_metrics: Metric = Metric(test_id=self.test_id)
+        test_metrics.time_taken = round(int(self.test.time_end - self.test.time_start))
+        test_metrics.time_start = datetime.fromtimestamp(self.test.time_start).strftime(DATE_TIME_TEMPLATE)
+        test_metrics.time_end = datetime.fromtimestamp(self.test.time_end).strftime(DATE_TIME_TEMPLATE)
+        test_metrics.success = self.test.success
+        with open(self.cgroup_path / 'memory.peak', 'r') as file:
+            test_metrics.memory_peak = file.read()
+
+        if (self.cgroup_path / 'cpu.stat').exists():
+            with open(self.cgroup_path / 'cpu.stat', 'r', ) as file:
+                self._parse_cpu_stat(file, test_metrics)
+        return test_metrics
+
+    def write_metrics_to_db(self, metrics: Metric, success: bool = False) -> None:
+        metrics.success = success
+        self.sqlite_writer.write_row(metrics, METRICS_TABLE)
+
+
+    def put_process_to_cgroup(self):
+        super().put_process_to_cgroup()
+        pid = os.getpid()
+        with open(self.cgroup_path / 'cgroup.procs', "a") as cgroup:
+            cgroup.write(str(pid))
+
+    def cgroup_monitor(self, test_event: Event) -> Task:
+        return asyncio.create_task(self._monitor_cgroup(test_event))
+
+    async def _monitor_cgroup(self, test_event: Event) -> None:
+        """Continuously monitors CPU and memory utilization."""
+        while not test_event.is_set():
+            with  open(self.cgroup_path / 'memory.current', 'r') as memory_current:
+                timeline_record = CgroupMetric(
+                    test_id=self.test_id,
+                    memory=int(memory_current.read()),
+                    timestamp=datetime.now()
+                )
+
+                self.sqlite_writer.write_row(timeline_record, CGROUP_MEMORY_METRICS_TABLE)
+
+            # Control the frequency of updates, for example, every 2 seconds
+            await asyncio.sleep(1)
+
+    @staticmethod
+    def _parse_cpu_stat(file: TextIO, metrics: Metric) -> None:
+        # Map the values from cpu.state to the model. Keys are values from cpu.stat
+        stats = {'user_usec': 'user_sec', 'system_usec': 'system_sec', 'usage_usec': 'usage_sec'}
+        for line in file.readlines():
+            stat, value = line.split(' ')
+            if stat in stats.keys():
+                setattr(metrics, stats[stat], float(value) / 1_000_000)
+
+
+def get_resource_gather(is_switched_on, test, tmpdir: str) -> ResourceGather:
```

Comment:
I'll add an annotation to the is_switched_on, but the test will be without it. Class Test is in test.py, and it's outside python module and test.py importing the resource_gather. If I'll try to add Test in resource_gather it will lead to a recursive import, and python will raise an exception.